Finish implementing Variant::Object and Variant::List

scovich opened the pull request:
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/7665

# Rationale for this change

Continuing the ongoing variant implementation effort.

# What changes are included in this PR?

As per title -- implement fairly complete support for variant objects and arrays. Also add some unit tests.

Note: This PR renames `VariantArray` as `VariantList` to align with parquet and arrow terminology, and to not conflict with the `VariantArray` we will eventually need to define for holding an arrow array of variant-typed data.

# Are there any user-facing changes?

Those variant subtypes should now be usable.

alamb commented:
I merged up from main to resolve a conflict with this branch

alamb reviewed (commented):
I made a PR to your branch that I think shows this code works well

- https://github.com/scovich/arrow-rs/pull/1

scovich commented on parquet-variant/src/variant.rs:
```diff
@@ -300,62 +303,218 @@ impl<'m> VariantMetadata<'m> {
+    pub fn field(&self, name: &str) -> Result<Option<Variant<'m, 'v>>, ArrowError> {
+        // Binary search through the sorted field IDs to find the field
+        let (field_ids, field_offsets) = self.parse_field_arrays()?;
+        let search_result = try_binary_search_by(&field_ids, &name, |&field_id| {
```
This also confused me at first... I probably should have added a code comment.

This binary search is over the field names of the variant object itself, which are indirectly referenced via the metadata dictionary. And the [spec](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md#value-data-for-object-basic_type2) does require those to be lexically ordered:
> The field ids and field offsets must be in lexicographical order of the corresponding field names in the metadata dictionary.

scovich replied:
(basically, if the requested field name actually exists, it must match the name referenced by one of the struct's field ids... and we can binary search them because those ids are in lexical order according to their backing dictionary entries)

scovich replied:
Aside: It would be incorrect to directly search the metadata dictionary, because that could "find" a field name that doesn't actually exist in the current object.

alamb commented on parquet-variant/src/variant.rs:
```diff
@@ -300,62 +303,218 @@ impl<'m> VariantMetadata<'m> {
-    pub fn get(&self, index: usize) -> Result<Variant<'m, 'v>, ArrowError> {
+    pub fn field(&self, name: &str) -> Result<Option<Variant<'m, 'v>>, ArrowError> {
+        // Binary search through the sorted field IDs to find the field
+        let (field_ids, field_offsets) = self.parse_field_arrays()?;
```
> I fully agree this is not an optimal API -- I just implemented the existing stub methods to give a starting point we can iterate on.

I think this is a good plan

alamb commented on parquet-variant/src/variant.rs:
```diff
@@ -300,62 +303,218 @@ impl<'m> VariantMetadata<'m> {
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct VariantListHeader {
```
Yup -- a follow on would be great

scovich replied:
Meanwhile, I made the new header structs and their methods `pub(crate)`. We can fix the others later once use cases are clearer.

alamb commented:
@scovich  -- what do you recommend for next steps? 
1. Shall I merge this PR and we can keep iterating on main?
2. WOuld you like to work on it some more?

I would like to get it merged in sooner rather than later to minimize conflicts such as https://github.com/apache/arrow-rs/pull/7670#pullrequestreview-2933021861

scovich commented on parquet-variant/src/variant.rs:
```diff
@@ -377,30 +534,134 @@ impl<'m, 'v> VariantArray<'m, 'v> {
+
+    /// Returns the offset size in bytes
+    pub fn offset_size(&self) -> usize {
+        self.offset_size as _
```
The `OffsetSizeBytes` enum occupies 1 byte in practice, so I don't think we'd save much? The offsets (2x `usize`) will anyway dwarf it.

scovich commented on parquet-variant/src/variant.rs:
```diff
@@ -717,4 +980,257 @@ mod tests {
             "unexpected error: {err:?}"
         );
     }
+
```
Do we actually need 256+ field names in the test? Or just verify our ability to process the wider offsets correctly?

scovich commented on parquet-variant/tests/variant_interop.rs:
```diff
@@ -99,7 +99,7 @@ fn variant_non_primitive() -> Result<(), ArrowError> {
             }
             "array_primitive" => match variant {
-                Variant::Array(arr) => {
+                Variant::List(arr) => {
```
Merged it in, thanks!

scovich commented:
> @scovich -- what do you recommend for next steps?
> 
>     1. Shall I merge this PR and we can keep iterating on main?
> 
>     2. WOuld you like to work on it some more?
> 
> I would like to get it merged in sooner rather than later to minimize conflicts such as [#7670 (review)](https://github.com/apache/arrow-rs/pull/7670#pullrequestreview-2933021861)

I addressed most of your review points and pushed; the only big outstanding question is how to handle the iterators. But we can probably tackle that in a separate PR?

alamb commented on parquet-variant/src/variant.rs:
```diff
@@ -377,30 +534,134 @@ impl<'m, 'v> VariantArray<'m, 'v> {
+        self.len() == 0
+    }
+
+    pub fn values(&self) -> Result<impl Iterator<Item = Variant<'m, 'v>>, ArrowError> {
```
Yeah, I think in general this is hitting the tension you mentioned in the earlier PRs of early vs late validation.

The more I think about it the more I think we should move the validation to construction because:
1. The only reason to create a `Variant` in the first place is to access its data, so I think we would end up validating it almost immediately on read
2. `Variants` are constructed once but read many times so validating up front is probably faster
3. We could offer an `unchecked` variant for construction if performance overhead is hight that skips the validation

alamb replied:
Filed two related issues
- https://github.com/apache/arrow-rs/issues/7685
- https://github.com/apache/arrow-rs/issues/7684

alamb commented on parquet-variant/src/variant.rs:
```diff
@@ -414,7 +415,11 @@ impl<'m, 'v> VariantObject<'m, 'v> {
-        // Binary search through the sorted field IDs to find the field
+        // Binary search through the field IDs of this object to find the requested field name.
+        //
+        // NOTE: This does not require a sorted metadata dictionary, because the variant spec
```
👍
